resources-service/migrations/20230223150400-create-resource-string-attributes-table.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (C) 2021-2023 Technology Matters
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize.query(`
+      CREATE TABLE IF NOT EXISTS resources."ResourceStringAttributes"
+      (
+        "resourceId" text COLLATE pg_catalog."default" NOT NULL,
+        "accountSid" text COLLATE pg_catalog."default" NOT NULL,
+        "key" text COLLATE pg_catalog."default" NOT NULL,
+        "value" text COLLATE pg_catalog."default" NOT NULL,
+        "language" text COLLATE pg_catalog."default" NOT NULL,
+        "info" JSONB,
+        "lastUpdated" timestamp with time zone,
+        "updateSequence"  bigint NOT NULL DEFAULT nextval('"Resources_updates_seq"'::regclass),
+        CONSTRAINT "ResourceStringAttributes_pkey" PRIMARY KEY ("resourceId", "accountSid", "key", "language", "value"),
```

Comment:
I'm still don't fully understand the need for improved indexing provided by splitting data into multiple tables by type that was described in the design doc. Could you outline a use case for those indexing improvements?

What is the thinking behind the complex primary key instead of using a sequential numeric `id` field and a UNIQUE index to enforce deduplication?

Does including `"value"` in the primary key break enforcement of deduplication?

Reply:
Nope... still don't grok the relationship. I'm moving back to my terraform world. No great reason to throw a monkey wrench into the things that you and Gian are already working towards and understand.

Reply:
Q1 - I'm not sure if improved indexing was the main driver behind the decision, TBH it was something of a coin flip, both options would work and both had their drawbacks. I just thought having compact tables that didn't expect most of their fields to be NULL would be better for storage, and from a data integrity perspective, there weren't corner cases around the 'type' column saying one thing but a different data column being populated set. Sharding the data by type might make lookups on the less used types quicker but I didn't have specific use cases around that.

FYI, unlike what's indicated in the proposal, now we are just having the primitive types with an additional JSONB column for each, it looks like we'd only need 4 tables for inline data, and I can't really see a compelling use case for reference values being anything other than strings, so we're only talking 3 extra tables in reality

Of course one other option not really discussed in the design proposal was the simplest one of all, don't bother with types, just 'stringly type' all the values :-P

Q2 - Just generally I prefer to use a natural PK, rather than creating a surrogate key and then adding a secondary unique constraint, especially if it aligns with a common query vector (like retrieving by resource ID). Also, the fact that sequential IDs being the only straightforward 'native' way to generate unique IDs within Postgres deters me from using surrogate keys if there is a natural alternative since they can cause maintenance headaches if we ever want to move data around from one db to another.  If UUID generation was natively supported I'd be more inclined to come down on the surrogate key side - I suppose we could install 'uuid-ossp' on our RDS instances since it's supported by RDS

Reply:
> Of course one other option not really discussed in the design proposal was the simplest one of all, don't bother with types, just 'stringly type' all the values :-P

Ha! This is exactly the direction my brain goes. Data comes in as json and leaves as json. Data types everywhere I can think of are restricted by the limitations of json. The added complexity of multiple tables gives me one of those "are we overcomplicating this?" gut reactions.

The driver behind the ID question was related to the trees question you asked in slack. In similar "un-opinionated" data storage systems I've built, I've used one-to-many relationships between data in a single table to create actual trees in the data. I've found that approach quite flexible in the long run. But I'll admit... I moved back to twilio-iac before I fully understood the current implementation.

I've never used postgres in anger from the developer side, only the infrastructure management side, and, even then, it has been several years...  so my understanding of postgres caveats is definitely limited. I'll circle back in a bit as I'm *finally* making some progress on twilio-iac refactor. :)

Reply:
So reading your Q, there is no direct relationship between ResourceStringAttributes and ResourceReferenceStringAttributeValues - they manage separate relationships to the Resources table ('inline' one to many values versus 'referenced' many to many value.

I've used the self joining table approach to build hierarchies in RDBMS systems in the past and the path approach that uses BTREE optimised LIKE queries to pull data. There's pros & cons to either. I've found the self joining could get a little unperformant as trees grew, but that was on older versions of SQL server, the use case has probably been optimised since, and I find there can be advantages to having the whole path back to root on each record without having to traverse back

I take your point about types, I suppose I wanted to keep the door open to performant filtering against dates & numbers, but there might not be much reason to do that directly on the Postgres DB (as opposed to via the search index) so it might be extra complexity for a purely theoretical use case

Sounds like there's nothing that means we need to slam on the breaks right now though, all the things we're discussing can be changed down the road so thx, I'll push on!

Reply:
Definitely nothing to hit the breaks over! Thanks for taking the time to write this stuff up, it is helpful to understand the thinking behind some of these choices.